packages/app-types/index.d.ts

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export type FirebaseOptions = {
+  apiKey?: string;
+  authDomain?: string;
+  databaseURL?: string;
+  projectId?: string;
+  storageBucket?: string;
+  messagingSenderId?: string;
+  [name: string]: any;
+};
+
+export class FirebaseApp {
+  constructor();
+  /**
+   * The (read-only) name (identifier) for this App. '[DEFAULT]' is the default
+   * App.
+   */
+  name: string;
+
+  /**
+   * The (read-only) configuration options from the app initialization.
+   */
+  options: FirebaseOptions;
+
+  /**
+   * Make the given App unusable and free resources.
+   */
+  delete(): Promise<void>;
+}
+
+export interface FirebaseNamespace {
+  /**
+   * Create (and intialize) a FirebaseApp.
+   *
+   * @param options Options to configure the services use in the App.
+   * @param name The optional name of the app to initialize ('[DEFAULT]' if
+   *   none)
+   */
+  initializeApp(options: FirebaseOptions, name?: string): FirebaseApp;
+
+  app: {
+    /**
+     * Retrieve an instance of a FirebaseApp.
+     *
+     * Usage: firebase.app()
+     *
+     * @param name The optional name of the app to return ('[DEFAULT]' if none)
+     */
+    (name?: string): FirebaseApp;
+
+    /**
+     * For testing FirebaseApp instances:
+     *  app() instanceof firebase.app.App
+     *
+     * DO NOT call this constuctor directly (use firebase.app() instead).
```

Comment:
This would be a change to how the other APIs function currently, lets keep this option on the table though.

Reply:
Calling the `PublicFirestore` constructor directly results in an Error being thrown, whereas the other components provide valid constructors to the firebase.<component>.<Component> path.

It's a way to prevent people from instantiating the class themselves, but it is a change in the current behavior of the SDK. We do specify in our documentation that the constructors (presumably valid) exist at that path even though we say not to use it.

Either way, if this is a change we really want to pursue, I'd say let's do it in follow up rather than in this PR.

Reply:
Okay, I'm fine with that.  I just think it's ugly to have a big scary warning in the comments / docs when we could just disallow it...  and since it's already documented as not supported, I wouldn't consider this a breaking change or anything. But I'm fine with punting on it for later.  Firestore already does it, which is all I really care about. :-)